agent/xds/clusters.go

```diff
@@ -547,43 +552,70 @@ func makeClusterFromUserConfig(configJSON string) (*envoy.Cluster, error) {
 	return &c, err
 }
 
-func (s *Server) makeGatewayCluster(cfgSnap *proxycfg.ConfigSnapshot, clusterName string) (*envoy.Cluster, error) {
-	return s.makeGatewayClusterWithConnectTimeout(cfgSnap, clusterName, 0)
-}
-
-// makeGatewayClusterWithConnectTimeout initializes a gateway cluster
-// with the specified connect timeout. If the timeout is 0, the connect timeout
-// defaults to use the configured gateway timeout.
-func (s *Server) makeGatewayClusterWithConnectTimeout(cfgSnap *proxycfg.ConfigSnapshot,
-	clusterName string, connectTimeout time.Duration) (*envoy.Cluster, error) {
+func (s *Server) makeGatewayCluster(snap *proxycfg.ConfigSnapshot,
+	name string, hostnameEndpoints structs.CheckServiceNodes, isRemote, onlyPassing bool, connectTimeout time.Duration) *envoy.Cluster {
 
-	cfg, err := ParseGatewayConfig(cfgSnap.Proxy.Config)
+	cfg, err := ParseGatewayConfig(snap.Proxy.Config)
 	if err != nil {
 		// Don't hard fail on a config typo, just warn. The parse func returns
 		// default config if there is an error so it's safe to continue.
 		s.Logger.Warn("failed to parse gateway config", "error", err)
 	}
-
 	if connectTimeout <= 0 {
 		connectTimeout = time.Duration(cfg.ConnectTimeoutMs) * time.Millisecond
 	}
 
-	cluster := envoy.Cluster{
-		Name:                 clusterName,
-		ConnectTimeout:       connectTimeout,
-		ClusterDiscoveryType: &envoy.Cluster_Type{Type: envoy.Cluster_EDS},
-		EdsClusterConfig: &envoy.Cluster_EdsClusterConfig{
+	cluster := &envoy.Cluster{
+		Name:           name,
+		ConnectTimeout: connectTimeout,
+
+		// Having an empty config enables outlier detection with default config.
+		OutlierDetection: &envoycluster.OutlierDetection{},
+	}
+
+	useEDS := true
+	if len(hostnameEndpoints) > 0 {
+		useEDS = false
+	}
+
+	// If none of the service instances are addressed by a hostname we provide the endpoint IP addresses via EDS
+	if useEDS {
+		cluster.ClusterDiscoveryType = &envoy.Cluster_Type{Type: envoy.Cluster_EDS}
+		cluster.EdsClusterConfig = &envoy.Cluster_EdsClusterConfig{
 			EdsConfig: &envoycore.ConfigSource{
 				ConfigSourceSpecifier: &envoycore.ConfigSource_Ads{
 					Ads: &envoycore.AggregatedConfigSource{},
 				},
 			},
-		},
-		// Having an empty config enables outlier detection with default config.
-		OutlierDetection: &envoycluster.OutlierDetection{},
+		}
+		return cluster
+	}
+
+	// When a service instance is addressed by a hostname we have Envoy do the DNS resolution
+	// by setting a DNS cluster type and passing the hostname endpoints via CDS.
+	rate := 10 * time.Second
+	cluster.DnsRefreshRate = &rate
+	cluster.DnsLookupFamily = envoy.Cluster_V4_ONLY
+
+	discoveryType := envoy.Cluster_Type{Type: envoy.Cluster_LOGICAL_DNS}
+	if cfg.DNSDiscoveryType == "strict_dns" {
+		discoveryType.Type = envoy.Cluster_STRICT_DNS
 	}
+	cluster.ClusterDiscoveryType = &discoveryType
 
-	return &cluster, nil
+	var endpoints []envoyendpoint.LbEndpoint
```

Comment:
```suggestion
	endpoints := make([]envoyendpoint.LbEndpoint, 0, len(hostnameEndpoints))
```
This is not likely to matter in practice, but it doesn't hurt.